File: app/src/main/java/opencrowd/hgc/hgcwallet/ui/main/pay/PayScreen.java
package opencrowd.hgc.hgcwallet.ui.main.pay;


import android.content.Context;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.text.Editable;
import android.text.TextWatcher;
import android.view.View;
import android.widget.Button;
import android.widget.CheckBox;
import android.widget.CompoundButton;
import android.widget.EditText;
import android.widget.LinearLayout;
import android.widget.RelativeLayout;
import android.widget.TextView;

import com.google.zxing.integration.android.IntentIntegrator;
import com.hederahashgraph.api.proto.java.Transaction;
import com.wealthfront.magellan.BaseScreenView;
import com.wealthfront.magellan.Screen;

import opencrowd.hgc.hgcwallet.App;
import opencrowd.hgc.hgcwallet.R;
import opencrowd.hgc.hgcwallet.hapi.APIRequestBuilder;
import opencrowd.hgc.hgcwallet.common.BaseTask;
import opencrowd.hgc.hgcwallet.common.TaskExecutor;
import opencrowd.hgc.hgcwallet.hapi.tasks.TransferTaskAPI;
import opencrowd.hgc.hgcwallet.app_intent.TransferRequestParams;
import opencrowd.hgc.hgcwallet.common.Singleton;
import opencrowd.hgc.hgcwallet.database.account.Account;
import opencrowd.hgc.hgcwallet.database.contact.Contact;
import opencrowd.hgc.hgcwallet.database.node.Node;
import opencrowd.hgc.hgcwallet.database.request.PayRequest;
import opencrowd.hgc.hgcwallet.modals.HGCAccountID;
import opencrowd.hgc.hgcwallet.ui.BaseActivity;
import opencrowd.hgc.hgcwallet.ui.customviews.TitleBarWrapper;
import opencrowd.hgc.hgcwallet.ui.main.MainActivity;
import opencrowd.hgc.hgcwallet.ui.main.pay.contacts.ContactListScreen;
import opencrowd.hgc.hgcwallet.ui.main.account.AccountListScreen;
import opencrowd.hgc.hgcwallet.ui.scan.QRScanListener;


public class PayScreen extends Screen<PayScreenView> implements AccountListScreen.AccountPickerListener,ContactListScreen.ContactPickerListener,QRScanListener{

    Account fromAccount;

    long fee = 0;
    @NonNull String toName = "", toAccIdStr = "", notes = "", amountStr = "";
    boolean addNote = false, showToAccount = false;

    public PayScreen(Account account) {
        this.fromAccount = account;
    }

    public PayScreen(Account fromAccount, PayRequest toAccount) {
        showToAccount = true;
        toAccIdStr = toAccount.getAccountId();
        toName = toAccount.getName();
        amountStr = Singleton.INSTANCE.toCoins(toAccount.getAmount()) + "";
        this.fromAccount = fromAccount;
    }

    @Override
    protected void onShow(Context context) {
        super.onShow(context);
        updateFee();
        getView().reloadData(this);
    }

    @NonNull
    @Override
    protected PayScreenView createView(@NonNull Context context) {
        return new PayScreenView(context, this);
    }

    @Override
    public void onAccountPick(Account account) {
        this.fromAccount = account;
        getNavigator().goBackTo(this);
    }

    @Override
    public void onContactPick(Contact contact) {
        showToAccount = true;
        toAccIdStr = contact.getAccountId();
        toName = contact.getName();
        getNavigator().goBackTo(this);
    }

    public void updateFee() {
        if (fromAccount.accountID() == null) return;

        double amount = 0;
        try {
            amount = Double.parseDouble(amountStr);
        } catch (Exception e) { }

        Transaction transaction = APIRequestBuilder.requestForTransfer(fromAccount, new HGCAccountID(0,0,0), Singleton.INSTANCE.toNanoCoins(amount), notes, APIRequestBuilder.PLACEHOLDE_FEE, Node.Companion.placeholderNode().accountID());
        fee = APIRequestBuilder.feeForTransferTransaction(transaction);

        PayScreenView view = getView();
        if (view != null) {
            view.updateFee(fee);
        }

    }

    public void onPayBuutonTap() {

        toAccIdStr = getView().mAccountIdEditText.getText().toString();
        toName = getView().mAccountNameEditText.getText().toString();
        amountStr = getView().mHgcAmountEditText.getText().toString();
        notes = getView().mNoteEditText.getText().toString();

        String error = null;
        HGCAccountID toAccountID = HGCAccountID.fromString(toAccIdStr);
        double amount = 0;

        try {
            amount = Double.parseDouble(amountStr);
        } catch (Exception e) { }

        if (toAccountID == null) {
            error = "Invalid to account id";
        } else if (amount == 0) {
            error = "Invalid amount";
        }
        final BaseActivity activity = (BaseActivity) getActivity();
        if (error == null) {
            activity.showActivityProgress("Please wait");
            final TransferTaskAPI task = new TransferTaskAPI(fromAccount, toAccountID, notes,toName, Singleton.INSTANCE.toNanoCoins(amount), fee);
            TaskExecutor taskExecutor = new TaskExecutor();
            taskExecutor.setListner(new TaskExecutor.TaskListner() {
                @Override
                public void onResult(BaseTask task1) {
                    activity.hideActivityProgress();
                    if (task.error == null) {
                        if (activity instanceof MainActivity) {
                            MainActivity mainActivity = (MainActivity)activity;
                            mainActivity.synchronizeData(true, false);
                        }
                        Singleton.INSTANCE.showToast(getActivity(), getActivity().getString(R.string.transaction_submitted_successfully));
                        getNavigator().goBack();

                    } else {
                        Singleton.INSTANCE.showToast(getActivity(), task.error);
                    }
                }
            });
            taskExecutor.execute(task);
        } else {
            Singleton.INSTANCE.showToast(getActivity(),error);
        }

    }

    @Override
    public void onQRScanFinished(boolean success, @Nullable String result) {
        if (!success) return;
        if (result != null) {
            TransferRequestParams params = TransferRequestParams.fromBarCode(result);
            if (params != null) {
                toAccIdStr = params.account.stringRepresentation();
                toName = params.name == null ? "" : params.name;
                amountStr = Singleton.INSTANCE.toCoins(params.amount) + "";
                if (params.note != null) {
                    notes = params.note;
                    addNote = true;
                }
                showToAccount = true;
                getView().reloadData(this);
            }
        }

    }
}

class PayScreenView extends BaseScreenView<PayScreen> {

    private TextView mAccountName;
    private TextView mPublicKey;
    private TextView mChangeText;
    private TextView mCancel;
    private TextView mNoteText, mFeeTextView;
    public EditText mNoteEditText, mHgcAmountEditText,mEditTextDollorAmount, mAccountIdEditText, mAccountNameEditText;

    private Button mNewButton,mScanButton,mExistingButton;
    private Button mPayButton;
    private CheckBox mAddNote;
    private LinearLayout mPayButtonsLayout,mNewButtonLayout,mScanButtonClickLayout;
    private RelativeLayout mAmountRelativeLayout;

    public PayScreenView(@NonNull final Context context, @NonNull PayScreen screen) {
        super(context);
        inflate(context, R.layout.view_pay_layout,this);
        mAccountName = (TextView)findViewById(R.id.text_account_name);
        mPublicKey = (TextView)findViewById(R.id.text_public_key);
        mChangeText = (TextView)findViewById(R.id.text_change);
        mCancel = (TextView)findViewById(R.id.text_cancel);
        mNewButton = (Button)findViewById(R.id.btn_new);
        mScanButton = (Button)findViewById(R.id.btn_scan);
        mPayButton = (Button)findViewById(R.id.btn_pay);
        mExistingButton = (Button)findViewById(R.id.btn_existing);
        mAddNote = (CheckBox) findViewById(R.id.add_note);
        mFeeTextView = findViewById(R.id.text_fee_value);

        //   mRequestNotification = (CheckBox)findViewById(R.id.request_notification);
        mNoteText = (TextView)findViewById(R.id.text_note);
        mNoteEditText = (EditText)findViewById(R.id.edittext_note);
        mHgcAmountEditText = (EditText)findViewById(R.id.edittext_hgc_amount);
        mEditTextDollorAmount = (EditText)findViewById(R.id.edittext_dollor_amount);
        mAccountIdEditText = (EditText)findViewById(R.id.edittext_account_id);
        mAccountNameEditText = (EditText)findViewById(R.id.edittext_account_name);

        mPayButtonsLayout= (LinearLayout)findViewById(R.id.pay_type_btn_layout);
        mNewButtonLayout = (LinearLayout)findViewById(R.id.new_btn_open_layout);
        mAmountRelativeLayout = (RelativeLayout)findViewById(R.id.amount_relative_layout);
        mHgcAmountEditText.addTextChangedListener(hgcEditTextListener);
        mEditTextDollorAmount.addTextChangedListener(dollerEditTextListener);

        reloadData(screen);
        mChangeText.setOnClickListener(new OnClickListener() {
            @Override
            public void onClick(View v) {
                getScreen().getNavigator().goTo(new AccountListScreen(getScreen()));
            }
        });

        mCancel.setOnClickListener(new OnClickListener() {
            @Override
            public void onClick(View v) {
                getScreen().showToAccount = false;
                setToAccountViewVisible(false);
            }
        });

        mExistingButton.setOnClickListener(new OnClickListener() {
            @Override
            public void onClick(View v) {
                getScreen().getNavigator().goTo(new ContactListScreen(getScreen()));
            }
        });

        mNewButton.setOnClickListener(new OnClickListener() {
            @Override
            public void onClick(View v) {
                getScreen().showToAccount = true;
                setToAccountViewVisible(true);
            }
        });

        mScanButton.setOnClickListener(new OnClickListener() {
            @Override
            public void onClick(View view) {
                new IntentIntegrator(getScreen().getActivity())
                        .setOrientationLocked(true)
                        .setPrompt("Place a QRcode inside the rectangle")
                        .initiateScan();
            }
        });



        mAddNote.setOnCheckedChangeListener(new CompoundButton.OnCheckedChangeListener() {
            @Override
            public void onCheckedChanged(CompoundButton buttonView, boolean isChecked) {
                getScreen().addNote = isChecked;
                setNoteTextFieldVisible(isChecked);
            }
        });

        mPayButton.setOnClickListener(new OnClickListener() {
            @Override
            public void onClick(View view) {
                getScreen().onPayBuutonTap();
            }
        });


        mHgcAmountEditText.setOnFocusChangeListener(new OnFocusChangeListener() {
            @Override
            public void onFocusChange(View view, boolean b) {
                if (!b){
                    getScreen().amountStr = mHgcAmountEditText.getText().toString();
                }
            }
        });

        mAccountNameEditText.setOnFocusChangeListener(new OnFocusChangeListener() {
            @Override
            public void onFocusChange(View view, boolean b) {
                if (!b){
                    getScreen().toName = mAccountNameEditText.getText().toString();
                }
            }
        });

        mAccountIdEditText.setOnFocusChangeListener(new OnFocusChangeListener() {
            @Override
            public void onFocusChange(View view, boolean b) {
                if (!b){
                    getScreen().toAccIdStr = mAccountIdEditText.getText().toString();
                }
            }
        });

        mNoteEditText.setOnFocusChangeListener(new OnFocusChangeListener() {
            @Override
            public void onFocusChange(View view, boolean b) {
                if (!b){
                    getScreen().notes = mNoteEditText.getText().toString();
                }
            }
        });
    }

    public void reloadData(@NonNull PayScreen screen) {
        Account fromAccount = screen.fromAccount;

        TitleBarWrapper titleBar = new TitleBarWrapper(findViewById(R.id.titleBar));
        titleBar.setTitle("PAY");
        titleBar.setCloseButtonHidden(false);

        titleBar.setOnCloseButtonClickListener(new OnClickListener() {
            @Override
            public void onClick(View v) {
                getScreen().getNavigator().goBack();
            }
        });

        mAccountName.setText(fromAccount.getName());

        if (fromAccount.accountID() != null) {
            mPublicKey.setText(fromAccount.accountID().stringRepresentation());
        } else {
            String shortKey = Singleton.INSTANCE.publicKeyStringShort(fromAccount);
            mPublicKey.setText(App.instance.getString(R.string.text_key_short,shortKey));
        }

        setText(mHgcAmountEditText, hgcEditTextListener, screen.amountStr);
        try {
            double coins = Double.parseDouble(screen.amountStr);
            double doller = Singleton.INSTANCE.hgcToUSD(Singleton.INSTANCE.toNanoCoins(coins));
            setText(mEditTextDollorAmount, dollerEditTextListener, Singleton.INSTANCE.toString(doller));
        } catch (Exception e) {

        }


        mNoteEditText.setText(screen.notes);
        mAccountIdEditText.setText(screen.toAccIdStr);
        mAccountNameEditText.setText(screen.toName);
        mAddNote.setChecked(screen.addNote);
        updateFee(screen.fee);
        setToAccountViewVisible(screen.showToAccount);
        setNoteTextFieldVisible(screen.addNote);
    }

    public void updateFee(long fee) {
        mFeeTextView.setText( Singleton.INSTANCE.formatHGC(fee, true));

    }

    public void setToAccountViewVisible(boolean visible) {
        if (visible) {
            mPayButtonsLayout.setVisibility(INVISIBLE);
            mNewButtonLayout.setVisibility(VISIBLE);
            RelativeLayout.LayoutParams p = new RelativeLayout.LayoutParams(LayoutParams.MATCH_PARENT,LayoutParams.WRAP_CONTENT);
            p.addRule(RelativeLayout.BELOW, R.id.new_btn_open_layout);
            p.setMargins(0, 100, 0, 0);
            mAmountRelativeLayout.setLayoutParams(p);

        } else  {
            mPayButtonsLayout.setVisibility(VISIBLE);
            mNewButtonLayout.setVisibility(INVISIBLE);
            RelativeLayout.LayoutParams p = new RelativeLayout.LayoutParams(LayoutParams.MATCH_PARENT,LayoutParams.WRAP_CONTENT);
            p.addRule(RelativeLayout.BELOW, R.id.pay_type_btn_layout);
            p.setMargins(0, 30, 0, 0);
            mAmountRelativeLayout.setLayoutParams(p);
        }
    }

    public void setNoteTextFieldVisible(boolean visible) {
        if(visible) {
            mNoteText.setVisibility(VISIBLE);
            mNoteEditText.setVisibility(VISIBLE);
        } else {
            mNoteText.setVisibility(GONE);
            mNoteEditText.setVisibility(GONE);
        }
    }

    private void setText(EditText editText, TextWatcher textWatcher, String text) {
        editText.removeTextChangedListener(textWatcher);
        editText.setText(text);
        editText.addTextChangedListener(textWatcher);
    }

    private TextWatcher dollerEditTextListener = new TextWatcher() {
        @Override
        public void beforeTextChanged(CharSequence charSequence, int i, int i1, int i2) {

        }

        @Override
        public void onTextChanged(CharSequence charSequence, int i, int i1, int i2) {

        }

        @Override
        public void afterTextChanged(Editable editable) {
            try {
                double doller = Double.parseDouble(editable.toString());
                double coins = Singleton.INSTANCE.toCoins(Singleton.INSTANCE.USDtoHGC(doller));
                setText(mHgcAmountEditText, hgcEditTextListener, Singleton.INSTANCE.toString(coins));

            } catch (Exception e) {

            }
        }
    };

    private TextWatcher hgcEditTextListener = new TextWatcher() {
        @Override
        public void beforeTextChanged(CharSequence charSequence, int i, int i1, int i2) {

        }

        @Override
        public void onTextChanged(CharSequence charSequence, int i, int i1, int i2) {

        }

        @Override
        public void afterTextChanged(Editable editable) {
            try {
                double coins = Double.parseDouble(editable.toString());
                double doller = Singleton.INSTANCE.hgcToUSD(Singleton.INSTANCE.toNanoCoins(coins));
                setText(mEditTextDollorAmount, dollerEditTextListener, Singleton.INSTANCE.toString(doller));


            } catch (Exception e) {

            }
        }
    };
}

File: app/src/main/java/hedera/hgc/hgcwallet/ui/main/pay/PayScreen.kt
/*
 *
 *  Copyright 2019 Hedera Hashgraph LLC
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 *
 */

package hedera.hgc.hgcwallet.ui.main.pay

import android.content.Context
import android.text.Editable
import android.text.TextWatcher
import android.view.View
import android.widget.*
import com.google.zxing.integration.android.IntentIntegrator
import com.squareup.okhttp.MediaType
import com.wealthfront.magellan.BaseScreenView
import com.wealthfront.magellan.Screen
import hedera.hgc.hgcwallet.*
import hedera.hgc.hgcwallet.app_intent.TransferRequestParams
import hedera.hgc.hgcwallet.common.Singleton
import hedera.hgc.hgcwallet.common.TaskExecutor
import hedera.hgc.hgcwallet.database.DBHelper
import hedera.hgc.hgcwallet.database.account.Account
import hedera.hgc.hgcwallet.database.contact.Contact
import hedera.hgc.hgcwallet.database.request.PayRequest
import hedera.hgc.hgcwallet.hapi.TransferParam
import hedera.hgc.hgcwallet.hapi.tasks.TransferTaskAPI
import hedera.hgc.hgcwallet.modals.ExchangeInfo
import hedera.hgc.hgcwallet.modals.HGCAccountID
import hedera.hgc.hgcwallet.ui.BaseActivity
import hedera.hgc.hgcwallet.ui.customviews.TitleBarWrapper
import hedera.hgc.hgcwallet.ui.main.MainActivity
import hedera.hgc.hgcwallet.ui.main.account.AccountListScreen
import hedera.hgc.hgcwallet.ui.main.pay.contacts.ContactListScreen
import hedera.hgc.hgcwallet.ui.scan.QRScanListener
import io.reactivex.Single
import io.reactivex.disposables.CompositeDisposable

class PayScreen(fromAccount: Account, type: PayType = PayType.Pay) : Screen<PayScreenView>(), AccountListScreen.AccountPickerListener, ContactListScreen.ContactPickerListener, QRScanListener {

    data class Params(var fromAccount: Account = Account(), val type: PayType = PayType.Pay, var showToAccount: Boolean = false, var addNote: Boolean = false, var toAccIdStr: String = "", var toName: String = "", var amountStr: String = "", var notes: String = "", var fee: Long = Singleton.getDefaultFee(), var host: String = "")

    private val param: Params
    private val disposables = CompositeDisposable()

    init {
        param = Params(fromAccount, type)
    }

    constructor(fromAccount: Account, toAccount: PayRequest, type: PayType) : this(fromAccount, type) {
        param.apply {
            showToAccount = true
            toAccIdStr = toAccount.accountId
            toName = toAccount.name ?: ""
            amountStr = Singleton.format(Singleton.toCoins(toAccount.amount))
        }

    }

    override fun onShow(context: Context?) {
        super.onShow(context)
        updateFee()
        view?.reloadData()
    }

    override fun createView(context: Context): PayScreenView {
        return PayScreenView(context, param)
    }

    override fun onAccountPick(account: Account) {
        param.fromAccount = account
        navigator.goBackTo(this)
    }

    override fun onContactPick(contact: Contact) {
        param.apply {
            showToAccount = true
            toAccIdStr = contact.accountId
            toName = contact.name ?: ""
            param.host = contact.getHost() ?: ""
        }
        navigator.goBackTo(this)
    }

    override fun onQRScanFinished(success: Boolean, result: String?) {
        if (!success) return
        if (result != null) {

            when (param.type) {
                PayType.Pay -> TransferRequestParams.fromBarCode(result)?.let { requestParams ->

                    param.apply {
                        toAccIdStr = requestParams.account.stringRepresentation()
                        toName = requestParams.name ?: ""
                        amountStr = Singleton.toCoins(requestParams.amount).toString() + ""
                        requestParams.note?.let {
                            notes = it
                            addNote = true
                        }
                        showToAccount = true
                    }
                    view.reloadData()
                }
                PayType.PayToExchange -> ExchangeInfo.fromQRCode(result)?.let { info ->
                    param.apply {
                        toAccIdStr = info.accountId.stringRepresentation()
                        toName = info.name
                        host = info.host
                        notes = info.memo ?: ""
                        showToAccount = true
                    }
                    view.reloadData()
                }
            }

        }
    }


    fun updateFee() {
        if (param.fromAccount.accountID() == null) return
        param.fee = Singleton.getDefaultFee()

        val view = view
        view?.updateFee(param.fee)
    }


    internal fun onPayButtonClick() {
        var error: String? = null
        val toAccountID = HGCAccountID.fromString(param.toAccIdStr)
        var amount = 0.0

        try {
            amount = java.lang.Double.parseDouble(param.amountStr)
        } catch (e: Exception) {
        }


        if (toAccountID == null) {
            error = "Invalid to account id"
        } else if (amount == 0.0) {
            error = "Invalid amount"
        }

        if (getByteCount(param.notes) > Config.maxAllowedMemoLength)
            error = "Note is too long"


        val activity = activity as BaseActivity
        if (error == null) {
            activity.showActivityProgress("Please wait")
            val transferParam = TransferParam(toAccountID!!, Singleton.toNanoCoins(amount), param.notes, param.fee, false, param.toName)
            val task = TransferTaskAPI(transferParam, param.fromAccount)
            val taskExecutor = TaskExecutor()
            taskExecutor.setListner {
                activity.hideActivityProgress()
                if (task.error == null) {
                    if (activity is MainActivity) {
                        activity.synchronizeData(true, false)
                    }
                    Singleton.showToast(activity, activity.getString(R.string.transaction_submitted_successfully))
                    navigator.goBack()

                } else {
                    task.error?.let { Singleton.showToast(activity, it) }
                }
            }
            taskExecutor.execute(task)
        } else {
            Singleton.showToast(getActivity(), error)
        }
    }

    internal fun onPayExchangeButtonClick() {
        var amount = 0.0
        try {
            amount = java.lang.Double.parseDouble(param.amountStr)
        } catch (e: Exception) {
        }

        val toAccountID = HGCAccountID.fromString(param.toAccIdStr)
        val MEDIA_TYPE_MARKDOWN = MediaType.parse("application/octet-stream")
        val hostUrl = ExchangeInfo.toHttpUrl(param.host)
        val activity = activity as BaseActivity

        var error = ""
        if (param.fromAccount.accountID() == null) {
            error = "Account is not linked"
        }

        if (toAccountID == null) {
            error = "Invalid to account id"
        } else if (amount == 0.0) {
            error = "Invalid amount"
        }

        if (getByteCount(param.notes) > Config.maxAllowedMemoLength)
            error = "Note is too long"

        if (error.isEmpty()) {
            activity.showActivityProgress("Please wait")
            val transferParam = TransferParam(toAccountID!!, Singleton.toNanoCoins(amount), param.notes, param.fee, forThirdParty(), param.toName)
            val txn = transferParam.getTransaction(param.fromAccount.getTransactionBuilder())
            disposables.add(
                    (+Single.fromCallable {
                        DBHelper.createContact(toAccountID, param.toName, param.host, true)
                    }).flatMap {
                        +postURL(hostUrl, txn.toByteArray())
                    }.flatMap {
                        Single.fromCallable {
                            DBHelper.createTransaction(txn, param.fromAccount.accountID())
                        }
                    }.subscribe({
                        activity.hideActivityProgress()
                        val success = true
                        if (success) {
                            if (activity is MainActivity) {
                                activity.synchronizeData(true, false)
                            }
                            Singleton.showToast(activity, activity.getString(R.string.transaction_submitted_successfully))
                            navigator.goBack()
                        } else
                            Singleton.showToast(getActivity(), error)

                    }, {
                        activity.hideActivityProgress()
                        Singleton.showToast(getActivity(), error)
                    })

            )
        } else
            Singleton.showToast(getActivity(), error)

    }

    fun forThirdParty() = when (param.type) {
        PayType.PayToExchange -> true
        PayType.Pay -> false
    }

    internal fun onChangeTextClick() {
        navigator?.goTo(AccountListScreen(this))
    }

    internal fun onCancelClick() {
        param.showToAccount = false
        view?.setToAccountViewVisible(false)
    }

    internal fun onExistingClick() {
        navigator?.goTo(ContactListScreen(this, forThirdParty()))
    }

    internal fun onNewContactClick() {
        param.showToAccount = true
        view?.setToAccountViewVisible(true)
    }

    internal fun onScanClick() {
        IntentIntegrator(activity)
                .setOrientationLocked(true)
                .setPrompt("Place a QRcode inside the rectangle")
                .initiateScan()
    }

    internal fun onAddNoteClick(isChecked: Boolean) {
        param.addNote = isChecked
        view?.setNoteTextFieldVisible(isChecked)
    }

    internal fun goBack() {
        navigator?.goBack()
    }

}

enum class PayType {
    PayToExchange,
    Pay
}

class PayScreenView(context: Context, val param: PayScreen.Params) : BaseScreenView<PayScreen>(context) {

    private val accountName: TextView?
    private val publicKey: TextView?
    private val changeText: TextView?
    private val cancel: TextView?
    private val feeTextView: TextView?
    private val feeTitleTextView: TextView?
    private val noteText: TextView?
    private val noteLayout: LinearLayout?
    private val newButton: Button?
    private val scanButton: Button?
    private val payButton: Button?
    private val existingButton: Button?
    private val addNote: CheckBox?
    private val noteEditText: EditText?
    private val hgcAmountEditText: EditText?
    private val editTextDollarAmount: EditText?
    private val accountIdEditText: EditText?
    private val hostEditText: EditText?
    private val accountNameEditText: EditText?
    private val payButtonsLayout: LinearLayout?
    private val newButtonLayout: LinearLayout?
    private val amountRelativeLayout: RelativeLayout?
    private val dollarAmountRelativeLayout: RelativeLayout?
    private val hgcAmountRelativeLayout: RelativeLayout?
    private val titleBarWrapper: TitleBarWrapper
    private var dollarAmountTextWatcher: TextWatcher? = null
    private var hgcAmountTextWatcher: TextWatcher? = null

    constructor(context: Context) : this(context, PayScreen.Params())

    init {
        View.inflate(context, R.layout.view_pay_layout, this)
        accountName = findViewById<TextView>(R.id.text_account_name)
        publicKey = findViewById(R.id.text_public_key)
        changeText = findViewById<TextView>(R.id.text_change)?.apply {
            setOnClickListener { screen?.onChangeTextClick() }
        }
        cancel = findViewById<TextView>(R.id.text_cancel)?.apply {
            setOnClickListener { screen?.onCancelClick() }
        }
        newButton = findViewById<Button>(R.id.btn_new)?.apply {
            setOnClickListener { screen?.onNewContactClick() }
        }
        scanButton = findViewById<Button>(R.id.btn_scan)?.apply {
            setOnClickListener { screen?.onScanClick() }
        }
        payButton = findViewById<Button>(R.id.btn_pay)?.apply {
            setOnClickListener {
                popluateData()
                when (param.type) {
                    PayType.Pay -> screen?.onPayButtonClick()
                    PayType.PayToExchange -> screen?.onPayExchangeButtonClick()
                }

            }
        }
        existingButton = findViewById<Button>(R.id.btn_existing)?.apply {
            setOnClickListener { screen?.onExistingClick() }
        }
        addNote = findViewById<CheckBox>(R.id.add_note)?.apply {
            setOnCheckedChangeListener { compoundButton, isChecked -> screen?.onAddNoteClick(isChecked) }
        }

        feeTitleTextView = findViewById(R.id.text_fee)
        feeTextView = findViewById(R.id.text_fee_value)

        noteLayout = findViewById(R.id.layout_note)
        noteText = findViewById(R.id.text_note)
        noteEditText = findViewById<EditText>(R.id.edittext_note)

        accountIdEditText = findViewById<EditText>(R.id.edittext_account_id)
        hostEditText = findViewById<EditText>(R.id.edittext_host)
        accountNameEditText = findViewById<EditText>(R.id.edittext_account_name)
        hgcAmountEditText = findViewById<EditText>(R.id.edittext_hgc_amount)
        editTextDollarAmount = findViewById<EditText>(R.id.edittext_dollor_amount)

        dollarAmountTextWatcher = object : TextWatcher {
            override fun afterTextChanged(editable: Editable?) {
                try {
                    val dollar = java.lang.Double.parseDouble(editable.toString())
                    val coins = Singleton.toCoins(Singleton.USDtoHGC(dollar))
                    hgcAmountTextWatcher?.let { setText(hgcAmountEditText, it, Singleton.toString(coins)) }
                } catch (e: Exception) {
                }

            }

            override fun beforeTextChanged(p0: CharSequence?, p1: Int, p2: Int, p3: Int) {}
            override fun onTextChanged(p0: CharSequence?, p1: Int, p2: Int, p3: Int) {}
        }

        hgcAmountTextWatcher = object : TextWatcher {
            override fun afterTextChanged(editable: Editable?) {
                try {
                    val coins = java.lang.Double.parseDouble(editable.toString())
                    val dollar = Singleton.hgcToUSD(Singleton.toNanoCoins(coins))
                    dollarAmountTextWatcher?.let { setText(editTextDollarAmount, it, Singleton.toString(dollar)) }
                } catch (e: Exception) {
                }
            }

            override fun beforeTextChanged(p0: CharSequence?, p1: Int, p2: Int, p3: Int) {}
            override fun onTextChanged(p0: CharSequence?, p1: Int, p2: Int, p3: Int) {}
        }

        editTextDollarAmount?.addTextChangedListener(dollarAmountTextWatcher)
        hgcAmountEditText?.addTextChangedListener(hgcAmountTextWatcher)


        payButtonsLayout = findViewById(R.id.pay_type_btn_layout)
        newButtonLayout = findViewById(R.id.new_btn_open_layout)
        amountRelativeLayout = findViewById(R.id.amount_relative_layout)
        dollarAmountRelativeLayout = findViewById(R.id.dollar_amount_layout)
        hgcAmountRelativeLayout = findViewById(R.id.hgc_amount_layout)


        titleBarWrapper = TitleBarWrapper(findViewById(R.id.titleBar)).apply {
            setCloseButtonHidden(false)
            setOnCloseButtonClickListener { screen?.goBack() }
        }


        when (param.type) {
            PayType.PayToExchange -> {
                dollarAmountRelativeLayout?.visibility = View.GONE
                (hgcAmountRelativeLayout?.getLayoutParams() as? (LinearLayout.LayoutParams))?.weight = 2f
                titleBarWrapper.setTitle("Pay to Third Party")
                addNote?.visibility = View.GONE
                feeTitleTextView?.visibility = View.GONE
                feeTextView?.visibility = View.GONE
                setNoteTextFieldVisible(false)
            }
            PayType.Pay -> {
                titleBarWrapper.setTitle("PAY")
            }
        }

        reloadData()
    }

    private fun popluateData() {
        accountIdEditText?.run {
            param.toAccIdStr = text.toString()
        }
        hostEditText?.run {
            param.host = text.toString()
        }

        accountNameEditText?.run {
            param.toName = text.toString()
        }
        hgcAmountEditText?.run {
            param.amountStr = text.toString()
        }
        noteEditText?.run {
            param.notes = text.toString()
        }
    }

    private fun setText(editText: EditText?, textWatcher: TextWatcher, text: String) {
        editText?.apply {
            removeTextChangedListener(textWatcher)
            setText(text)
            addTextChangedListener(textWatcher)
        }
    }

    fun reloadData() {
        accountName?.setText(param.fromAccount.name)

        publicKey?.text = if (param.fromAccount.accountID() != null) {
            param.fromAccount.accountID()!!.stringRepresentation()
        } else {
            val shortKey = Singleton.publicKeyStringShort(param.fromAccount)
            context.getString(R.string.text_key_short, shortKey)
        }

        hgcAmountTextWatcher?.let { setText(hgcAmountEditText, it, param.amountStr) }

        try {
            val coins = java.lang.Double.parseDouble(param.amountStr)
            val dollar = Singleton.hgcToUSD(Singleton.toNanoCoins(coins))
            dollarAmountTextWatcher?.let { setText(editTextDollarAmount, it, Singleton.toString(dollar)) }

        } catch (e: Exception) {

        }

        noteEditText?.setText(param.notes)
        accountIdEditText?.setText(param.toAccIdStr)
        accountNameEditText?.setText(param.toName)
        hostEditText?.setText(param.host)
        addNote?.isChecked = param.addNote
        updateFee(param.fee)
        setToAccountViewVisible(param.showToAccount)
        setNoteTextFieldVisible(param.addNote)
    }

    fun updateFee(fee: Long) {
        feeTextView?.setText(Singleton.formatHGC(fee, true))
    }


    fun setToAccountViewVisible(visible: Boolean) {
        if (visible) {
            payButtonsLayout?.setVisibility(View.INVISIBLE)
            newButtonLayout?.setVisibility(View.VISIBLE)
            hostEditText?.visibility = when (param.type) {
                PayType.PayToExchange -> View.VISIBLE
                else -> View.GONE
            }
            val p = RelativeLayout.LayoutParams(FrameLayout.LayoutParams.MATCH_PARENT, FrameLayout.LayoutParams.WRAP_CONTENT)
            p.addRule(RelativeLayout.BELOW, R.id.new_btn_open_layout)
            p.setMargins(0, 100, 0, 0)
            amountRelativeLayout?.setLayoutParams(p)

        } else {
            payButtonsLayout?.setVisibility(View.VISIBLE)
            newButtonLayout?.setVisibility(View.INVISIBLE)
            val p = RelativeLayout.LayoutParams(FrameLayout.LayoutParams.MATCH_PARENT, FrameLayout.LayoutParams.WRAP_CONTENT)
            p.addRule(RelativeLayout.BELOW, R.id.pay_type_btn_layout)
            p.setMargins(0, 30, 0, 0)
            amountRelativeLayout?.setLayoutParams(p)
        }
    }

    fun setNoteTextFieldVisible(visible: Boolean) {
        val visibility = if (visible) View.VISIBLE else View.GONE
        noteLayout?.visibility = visibility
    }


}